### PR TITLE
Users/saprajap/resend failed notifications

### PR DESCRIPTION
**Pull request by sanjayprajap**
Resend by dataRange feature added. This api call will trigger the falied notification in given daterange on notification field 'SendOnUTCDate'.

### CODE REVIEW

File: NotificationService/NotificationHandler/appsettings.json [hunk 1]
```diff
@@ -67,5 +67,7 @@
     "MailTemplateTableName": "__MailTemplateTableName__"
   },
   "StorageType": "__StorageType__",
-  "NotificationProviderType": "__NotificationProviderType__"
+  "NotificationProviderType": "__NotificationProviderType__",
+  "KeyVaultConfigRefreshDurationSeconds": "__KeyVaultConfigRefreshDurationSeconds__",
+  "AllowedMaxResendDays": "__AllowedMaxResendDays__"
```
Comment (reddaiahnethi) [inline NotificationService/NotificationHandler/appsettings.json:72]: give this new setting to Deepika

Comment (reddaiahnethi) [reply NotificationService/NotificationHandler/appsettings.json:72]: AllowedMaxResendDurationInDays

File: NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailHandlerManager.cs [hunk 2]
```diff
@@ -234,7 +234,7 @@ public async Task<IList<NotificationResponse>> ResendEmailNotifications(string a
             IList<NotificationResponse> notificationResponses = new List<NotificationResponse>();
 
             // Queue a single cloud message for all entities created to enable parallel processing.
-            var cloudQueue = this.cloudStorageClient.GetCloudQueue("notifications-queue");
+            var cloudQueue = this.cloudStorageClient.GetCloudQueue(Constants.NotificationsQueue);
```
Comment (reddaiahnethi) [inline NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailHandlerManager.cs:237]: Read it from Config. Why constant?

File: NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailHandlerManager.cs [hunk 3]
```diff
@@ -249,5 +249,27 @@ public async Task<IList<NotificationResponse>> ResendEmailNotifications(string a
...
+        public async Task<IList<NotificationResponse>> ResendEmailNotificationsByDateRange(string applicationName, DateTimeRange dateRange)
+        {
+            this.logger.TraceInformation($"Started {nameof(this.ResendEmailNotificationsByDateRange)} method of {nameof(EmailHandlerManager)}.");
+            var maxAllowedDaysInRange = (double)this.configuration.GetValue(typeof(double), Constants.AllowedMaxResendDays);
+            if (dateRange != null && (dateRange.EndDate - dateRange.StartDate).TotalDays >= maxAllowedDaysInRange)
+            {
+                throw new DataException($"Date-range must not be less or equal to {maxAllowedDaysInRange}");
+            }
+
+            var failedNtificationEntities = await this.emailManager.GetEmailNotificationsByDateRange(applicationName, dateRange).ConfigureAwait(false);
```
Comment (reddaiahnethi) [inline NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailHandlerManager.cs:263]: GetPendingOrFailedEmailNotificationByDateTange

Comment (reddaiahnethi) [inline NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailHandlerManager.cs:263]: Spell check

Comment (reddaiahnethi) [reply NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailHandlerManager.cs:263]: failedNtificationEntities

File: NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailHandlerManager.cs [hunk 4]
```diff
@@ -249,5 +249,27 @@ public async Task<IList<NotificationResponse>> ResendEmailNotifications(string a
...
+            if (failedNtificationEntities == null || failedNtificationEntities.Count == 0)
+            {
+                return null;
+            }
+
+            var notificationIds = failedNtificationEntities.Select(notificationEntity => notificationEntity.NotificationId);
+            var result = await this.ResendEmailNotifications(applicationName, notificationIds.ToArray()).ConfigureAwait(false);
```
Comment (reddaiahnethi) [inline NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailHandlerManager.cs:270]: Here, this should ignore the sent items. If api call timeout, and user hits again, it should not send again. So, Make sure you IgnoreAlreadyent items to true

**Review by deepika791 — APPROVED** [review]

File: NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailManager.cs [hunk 5]
```diff
@@ -315,5 +315,12 @@ public async Task<IList<MeetingNotificationItemEntity>> CreateMeetingNotificatio
             this.logger.TraceInformation($"Completed {nameof(this.CreateNotificationEntities)} method of {nameof(EmailManager)}.", traceProps);
             return notificationEntities;
         }
+
+        /// <inheritdoc/>
+        public async Task<IList<EmailNotificationItemEntity>> GetEmailNotificationsByDateRange(string applicationName, DateTimeRange dateRange)
```
Comment (reddaiahnethi) [inline NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailManager.cs:320]: Change these name, as you are getting only failed. If you want to make this method generic, accept filters as parameter and add status=failed as filter

File: NotificationService/NotificationService.Common/Constants.cs [hunk 6]
```diff
@@ -213,5 +213,14 @@ public static class Constants
...
 
+        /// <summary>
+        /// A constant for ResendDateRange logging.
+        /// </summary>
+        public const string ResendDateRange = "ResendDateRange";
+
+        /// <summary>
+        /// A constant or resend allowed days.
+        /// </summary>
+        public const string AllowedMaxResendDays = "AllowedMaxResendDays";
```
Comment (reddaiahnethi) [inline NotificationService/NotificationService.Common/Constants.cs:224]: Change the name AllowedMaxResendDurationInDays

**Review by reddaiahnethi — CHANGES_REQUESTED** [review]

File: NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailHandlerManager.cs (hunk 4 above)
Comment (sanjayprajap) [reply NotificationService/NotificationService.BusinessLibrary/Business/v1/EmailHandlerManager.cs]: changes done as suggested.

**Review by deepika791 — APPROVED** [review]

**Review by reddaiahnethi — APPROVED** [review]

**Review by deepika791 — APPROVED** [review]

**Pull request merged**